web/src/message_report.ts

```diff
@@ -31,13 +34,99 @@ export const message_report_type_options: Option[] = [
     },
 ];
 
+type MessagePreviewRenderContext = {
+    include_sender: boolean;
+    msg: Message;
+    sender_is_bot: boolean;
+    should_add_guest_indicator_for_sender: boolean;
+    small_avatar_url: string;
+    timestr: string;
+    message_list_id: string;
+    status_message: boolean;
+    want_date_divider: boolean;
+    hide_message_reactions: boolean;
+    hide_message_controls: boolean;
+};
+
+function register_message_preview_click_handlers(
+    message_preview_container: JQuery,
+    sender_id: number,
+): void {
+    // This function registers click handlers and mouseover effects
+    // for the message sender in the message preview container.
+    // The logic here is partly from message_list_hover.ts, and
+    // partly from user_card_popover.ts.
+
+    message_preview_container.on("mouseover", ".sender_info_hover", function (this: HTMLElement) {
+        const $row = $(this).closest(".message_row");
+        $row.addClass("sender_info_hovered");
+    });
+
+    message_preview_container.on("mouseout", ".sender_info_hover", function (this: HTMLElement) {
+        const $row = $(this).closest(".message_row");
+        $row.removeClass("sender_info_hovered");
+    });
+
+    message_preview_container.on(
+        "click",
+        ".sender_name, .inline-profile-picture-wrapper",
+        function (this: HTMLElement, e) {
+            e.stopPropagation();
+            const user = people.get_by_user_id(sender_id);
+            toggle_user_card_popover_for_message(this, user, sender_id, true);
+        },
+    );
+}
+
+function get_message_group_for_message_preview(message: Message): MessageGroup {
+    // This creates a simpler recipient_row without element like the
+    // topic menu and topic visibility policy menu.
+    const message_group = populate_group_from_message(message, false, false, undefined, true, true);
+    if (message_group.is_stream) {
+        message_group.user_can_resolve_topic = false;
+        message_group.is_topic_editable = false;
+    }
+    return message_group;
+}
+
+function get_message_preview_body_args(message: Message): MessagePreviewRenderContext {
+    // This creates a simpler message body without elements like the
+    // message actions and reactions.
+    const include_sender = true;
+    const sender_is_bot = people.sender_is_bot(message);
+    const should_add_guest_indicator_for_sender = people.should_add_guest_user_indicator(
+        message.sender_id,
+    );
+    const small_avatar_url = people.small_avatar_url(message);
+    const timestr = get_timestr(message);
+    return {
+        include_sender,
+        msg: message,
+        sender_is_bot,
+        should_add_guest_indicator_for_sender,
+        small_avatar_url,
+        timestr,
+        message_list_id: "",
+        status_message: false,
+        want_date_divider: false,
+        hide_message_reactions: true,
+        hide_message_controls: true,
+    };
+}
```

Comment:
How are you validating that this has all the properties required to render the template? 

I feel like it would be a lot safer if you were generating a type like `MessageContainer` or `MessageGroup` so that the type-checker will notice if someone extends those types without changing this code path.